cmd/drop.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+module Homebrew
+    module_function
+
+    def drop_args
+        Homebrew::CLI::Parser.new do
+            description <<~EOS
+                Drop one or more <formula>e and/or <cask>s from a `Brewfile`.
+
+                You can specify the `Brewfile` location using `--file`, `--global`,
+                or by setting the `HOMEBREW_BUNDLE_FILE` environment variable.
+            EOS
+
+            switch "-g", "--global",
+                description: "Read the `Brewfile` from `~/.Brewfile`."
+            flag "--file=",
+                description: "Read the `Brewfile` from this location. Use `--file=-` to pipe to stdin/stdout."
+
+            conflicts "global", "file"
+
+            switch "--formula", "--formulae", description: "Treat all named arguments as formulae."
+            switch "--cask", "--casks", description: "Treat all named arguments as casks."
+
+            conflicts "formula", "cask"
+
+            named_args [:formula, :cask], min: 1
+        end
+    end
+
+    def drop
+        args = drop_args.parse
+        silent = args.quiet?
+
+        brews = args.named.to_formulae_and_casks
+
+        # Require this file: https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle.rb
+        require_relative "../../../homebrew/homebrew-bundle/lib/bundle"
+
+        brewfile = Bundle::Brewfile.path(global: args.global?, file: args.file)
+
+        ohai "Using Brewfile at '#{brewfile}'" unless silent
+
+        # Parse the Brewfile and create separate arrays for formulae, casks, and taps.
+        parsed_entries = Bundle::Dsl.new(brewfile.read).entries
+        bundle_taps, bundle_formulae, bundle_casks = [], [], []
+        parsed_entries.each do |entry|
+            case entry.type
+                when :tap; bundle_taps
+                when :brew; bundle_formulae
+                when :cask; bundle_casks
+                else; next # Matches `mas` and `whalebrew`.
+            end << entry.name
+        end
+
+        # Open the Brewfile and drop the requested items from it.
+        brews.each do |brew|
+
+            # Get the corresponding reference list and terms for the given item.
+            current_bundle_list, brewfile_prefix_type, display_type = if brew.is_a?(Formula)
+                [bundle_formulae, "brew", "Formula"]
+            else
+                [bundle_casks, "cask", "Cask"]
+            end
+
+            # Drop the formula/cask from the file if it exists.
+            unless current_bundle_list.include?(brew.full_name) || current_bundle_list.include?(brew.name)
+                opoo "#{display_type} '#{brew.full_name}' is already not present in the Brewfile. Skipping." unless silent
+            else
+                lines = []
+
+                File.foreach("Brewfile") do |line|
+                    if not line.match(/#{brewfile_prefix_type} ["']#{brew.full_name}["']/)
```

Comment:
This isn't relevant to the PR, but I think that removing lines from a Brewfile by skipping over them when they are written back to the file is a very smart way to do it and very elegant. Good job!

First of all, a minor nitpick. You can use `unless` instead of `if not`:
```suggestion
                    unless line.match(/#{brewfile_prefix_type} ["']#{brew.full_name}["']/)
```

---

Also, I'm worried about how restrictive this line is. It will match on more complex checks like these, which is good:
```
 cask "vscodium", args: { no_quarantine: true, require_sha: true } # Open Source Code Editor
 ```

But it only checks the `brew.full_name`, which means that for formulae/casks that come from third-party taps, the full name might not match. For example, I can't run `brew drop xshe` because it appears in the `Brewfile` as `brew "xshe"` alongside a corresponding `tap` declaration, as opposed to the full form of `brew "superatomic/xshe/xshe"`.
```suggestion
                    if not line.match(/#{brewfile_prefix_type} ["'](#{brew.full_name}|#{brew.name})["']/)
```
This suggestion resolves this problem, although it will causes problems if two formulae/casks share an identical `brew.name`. This might require an additional check. Let me know how you feel about what to do here.